src/script/sign.cpp

```diff
@@ -509,6 +509,9 @@ bool SignTransaction(CMutableTransaction& mtx, const SigningProvider* keystore,
             } else {
                 input_errors[i] = ScriptErrorString(serror);
             }
+        } else {
+            // If this input succeeds, make sure there is no error set for it
+            input_errors.erase(i);
```

Comment:
Help me understand:

1) Two SPKM: A and B
2) Two input tx:
input 1: key resides in A
input 2: key resides in B
3) A asked to sign 1, successfully
4) A asked to sign 2, unsuccessful, appends error C
5) B asked to sign 1, unsuccessful, appends error D
6) B asked to sign 2, successful, deletes C

We're left with error D?

Reply:
5. B asked to sign 1. 1 comes with existing signature. Skips signing because already signed, no error.

Reply:
`CWallet::SignTransaction` outer loop over inputs has the complete check, not the inner SPKM loop?

Reply:
@achow101 `CWallet` skips B when considering input 1, because it's already signed. But then when it's considering input 2 it will tell B to sign the transaction, and the sign code doesn't seem skip already signed inputs.

It should be easy to reproduce this scenario in a test on the descriptor wallet PR.

Reply:
To be more specific, the generic `SignTransaction` function uses `DataFromTransaction` on each input to pull out any existing signatures, pubkeys, and scripts into a `SignatureData` object. When this `SignatureData` object gets passed into `ProduceSignature`, `ProduceSignature` first looks at the `SignatureData` finds an existing signatures, pubkeys, and scripts, and fills those into the transaction rather than asking for the `SigningProvider` (which is produced by the `ScriptPubKeyMan`). So even if that `SigningProvider` can't sign, because we already had signatures, pubkeys, and scripts in that transaction, no error occurs.

There is an additional optimization already in `ProduceSignature` which skips signing when it finds that `SignatureData.complete == true`, which is determined by doing a `VerifyScript` within `DataFromTransaction`. Hence signing is skipped for that input.

All of this is pre-existing code which allows us to sign a transaction with external inputs without error. The same process that we do internally with multiple `ScriptPubKeyMan`s is almost the same as if you signed a transaction that has inputs from multiple wallets with those different wallets. Errors are not given for the inputs that are already signed.

Reply:
Ok, I agree with what you're saying, though it's indirect.

Reply:
That makes sense and will do for now. Would be good to add a test to #15764 for this type of stuff, because it seems brittle.

Reply:
The nature of descriptor wallets as implemented in #16528 requires this commit or tests fail. We are already testing this case. I added back this commit *because* tests were failing without it in #16528.